cleverhans/future/torch/attacks/deepfool.py

```diff
@@ -0,0 +1,175 @@
+      logits,
+    )
+    target_classes = incorrect_logits.argsort(
+      dim=1, descending=True)[:, :num_classes]
+
+  x = x.clone().detach().to(torch.float)
+  perturbations = torch.zeros_like(x)
+
+  if is_debug:
+    print("Starting DeepFool attack")
+
+  for i in range(max_iter):
+    x_adv = x + (1 + overshoot) * perturbations
+    x_adv.requires_grad_(True)
+    zero_gradients(x_adv)
+    logits = model_fn(x_adv)
+
+    # "Live" inputs are still being attacked; others have already achieved misclassification
+    if targeted:
+      live = torch.argmax(logits, dim=1) != target_classes[:, 0]
+    else:
+      live = torch.argmax(logits, dim=1) == y
+    if is_debug:
+      print('Iteration {}: {:.1f}% success'.format(
+        i, 100 * (1 - live.sum().float() / len(live)).item()))
+    if torch.all(~live):
+      # Stop early if all inputs are already misclassified
+      break
+
+    smallest_magnitudes = torch.full((int(live.sum()),), np.inf,
+                                     dtype=torch.float, device=perturbations.device)
+    smallest_perturbation_updates = torch.zeros_like(perturbations[live])
+
+    logits[live, y[live]].sum().backward(retain_graph=True)
+    grads_correct = x_adv.grad.data[live].clone().detach()
+
+    for k in range(target_classes.size()[1]):
+      zero_gradients(x_adv)
+
+      logits_target = logits[live, target_classes[live, k]]
+      logits_target.sum().backward(retain_graph=True)
+      grads_target = x_adv.grad.data[live].clone().detach()
+
+      grads_diff = (grads_target - grads_correct).detach()
+      logits_margin = (logits_target - logits[live, y[live]]).detach()
+
+      p = 1 if norm == np.inf else 2
+
+      grads_norm = (grads_diff ** p).abs().sum(dim=list(range(1, len(grads_diff.size())))) \
+        ** (1. / p)
+      magnitudes = logits_margin.abs() / grads_norm
+
+      magnitudes_expanded = magnitudes
+      for _ in range(len(grads_diff.size()) - 1):
+        grads_norm = grads_norm.unsqueeze(-1)
+        magnitudes_expanded = magnitudes_expanded.unsqueeze(-1)
+
+      if norm == np.inf:
+        perturbation_updates = ((magnitudes_expanded + 1e-4) *
+                                torch.sign(grads_diff))
+      else:
+        perturbation_updates = ((magnitudes_expanded + 1e-4) * grads_diff /
+                                 grads_norm)
+
+      smaller = magnitudes < smallest_magnitudes
+      smallest_perturbation_updates[smaller] = perturbation_updates[smaller]
+      smallest_magnitudes[smaller] = magnitudes[smaller]
+
+    all_perturbation_updates = torch.zeros_like(perturbations)
+    all_perturbation_updates[live] = smallest_perturbation_updates
+    perturbations.add_(all_perturbation_updates)
+
+    perturbations *= (1 + overshoot)
+    if eps is not None:
+      perturbations = clip_eta(perturbations, norm, eps)
+    perturbations = torch.clamp(x + perturbations, clip_min, clip_max) - x
+    perturbations /= (1 + overshoot)
+
+#   perturbations *= (1 + overshoot)
```

Comment:
Why is this code commented?

Reply:
Sorry, that shouldn't be there actually. I'll remove it